### ISSUE TITLE
Loop action not performed in interactive mode (4.3)

### ISSUE THREAD

**Issue by AdamWill** [os-autoinst/os-autoinst#400]
I was re-doing Fedora needles for a GTK+ / Cantarell change today, first time doing it with openQA / os-autoinst 4.3. I noticed what I think is a new bug. We have a test with a loop like this, for handling a software selection screen:

    # select desired environment
    # go through the list 20 times at max (to prevent infinite loop when it's missing)
    for (my $i = 0; !check_screen("anaconda_".$packageset."_highlighted", 1) && $i < 20; $i++) {
               send_key "down";
    }

i.e. a loop which checks if a needle matches, if it doesn't, presses 'down', then checks again. When running in interactive mode, it seems as if the `send_key` never happens when I click the move forward ('play') button - the needle comparison does, but the key event doesn't. So I can never get the test to actually highlight the appropriate list entry so I can edit the needle. Here's an os-autoinst log extract. Note that it starts out in non-interactive mode - I was trying to switch to interactive mode exactly when it hit the right menu entry, but didn't quite make it. So you see it hit 'tab' twice (which is the thing it does right before the loop), then it starts running through the loop as normal in non-interactive mode, but when I switch interactive mode on, you can see that each time I continue, it checks the needle but doesn't do the `send_key`. When I switch interactive mode back off, it goes back to working normally.

    Debug: /var/lib/openqa/share/tests/fedora/tests/_software_selection.pm:18 called testapi::send_key
    <<< send_key(key='tab')
    Debug: /var/lib/openqa/share/tests/fedora/tests/_software_selection.pm:20 called testapi::send_key
    <<< send_key(key='tab')
    Debug: /var/lib/openqa/share/tests/fedora/tests/_software_selection.pm:24 called testapi::check_screen
    <<< check_screen(mustmatch='anaconda_kde_highlighted', timeout=1)
    MATCH(anaconda_kde_highlighted:0.00)
    MATCH(anaconda_kde_highlighted-freetype262:0.00)
    no match 1
    MATCH(anaconda_kde_highlighted:0.00)
    MATCH(anaconda_kde_highlighted-freetype262:0.00)
    no match 0
    MATCH(anaconda_kde_highlighted:0.00)
    MATCH(anaconda_kde_highlighted-freetype262:0.00)
    >>> assert_screen: match=anaconda_kde_highlighted timed out after 1
    Debug: /var/lib/openqa/share/tests/fedora/tests/_software_selection.pm:24 called testapi::send_key
    <<< send_key(key='down')
    Debug: /var/lib/openqa/share/tests/fedora/tests/_software_selection.pm:24 called testapi::check_screen
    <<< check_screen(mustmatch='anaconda_kde_highlighted', timeout=1)
    MATCH(anaconda_kde_highlighted:0.00)
    MATCH(anaconda_kde_highlighted-freetype262:0.00)
    no match 1
    MATCH(anaconda_kde_highlighted:0.00)
    MATCH(anaconda_kde_highlighted-freetype262:0.00)
    no match 0
    MATCH(anaconda_kde_highlighted:0.00)
    MATCH(anaconda_kde_highlighted-freetype262:0.00)
    >>> assert_screen: match=anaconda_kde_highlighted timed out after 1
    Debug: /var/lib/openqa/share/tests/fedora/tests/_software_selection.pm:24 called testapi::send_key
    <<< send_key(key='down')
    Debug: /var/lib/openqa/share/tests/fedora/tests/_software_selection.pm:24 called testapi::check_screen
    <<< check_screen(mustmatch='anaconda_kde_highlighted', timeout=1)
    MATCH(anaconda_kde_highlighted:0.00)
    MATCH(anaconda_kde_highlighted-freetype262:0.00)
    no match 1
    MATCH(anaconda_kde_highlighted:0.00)
    MATCH(anaconda_kde_highlighted-freetype262:0.00)
    no match 0
    MATCH(anaconda_kde_highlighted:0.00)
    MATCH(anaconda_kde_highlighted-freetype262:0.00)
    EVENT {"data":{"path":"/machine/peripheral-anon/device[0]/virtio-backend"},"event":"NIC_RX_FILTER_CHANGED","timestamp":{"seconds":1453407266,"microseconds":165273}}
    EVENT {"event":"STOP","timestamp":{"seconds":1453407332,"microseconds":824979}}
    interactive mode waiting for continuation
    continuing
    EVENT {"event":"RESUME","timestamp":{"seconds":1453407336,"microseconds":900928}}
    MATCH(anaconda_kde_highlighted:0.00)
    MATCH(anaconda_kde_highlighted-freetype262:0.00)
    no match 5
    no change 4
    no change 3
    no change 2
    no change 1
    no change 0
    MATCH(anaconda_kde_highlighted:0.00)
    MATCH(anaconda_kde_highlighted-freetype262:0.00)
    EVENT {"event":"STOP","timestamp":{"seconds":1453407343,"microseconds":27644}}
    interactive mode waiting for continuation
    continuing
    EVENT {"event":"RESUME","timestamp":{"seconds":1453407374,"microseconds":110328}}
    MATCH(anaconda_kde_highlighted:0.00)
    MATCH(anaconda_kde_highlighted-freetype262:0.00)
    no match 5
    no change 4
    no change 3
    no change 2
    no change 1
    no change 0
    MATCH(anaconda_kde_highlighted:0.00)
    MATCH(anaconda_kde_highlighted-freetype262:0.00)
    EVENT {"event":"STOP","timestamp":{"seconds":1453407380,"microseconds":233687}}
    interactive mode waiting for continuation
    continuing
    EVENT {"event":"RESUME","timestamp":{"seconds":1453407391,"microseconds":311140}}
    MATCH(anaconda_kde_highlighted:0.00)
    MATCH(anaconda_kde_highlighted-freetype262:0.00)
    no match 5
    no change 4
    no change 3
    no change 2
    no change 1
    no change 0
    MATCH(anaconda_kde_highlighted:0.00)
    MATCH(anaconda_kde_highlighted-freetype262:0.00)
    >>> assert_screen: match=anaconda_kde_highlighted timed out after 1
    Debug: /var/lib/openqa/share/tests/fedora/tests/_software_selection.pm:24 called testapi::send_key
    <<< send_key(key='down')
    Debug: /var/lib/openqa/share/tests/fedora/tests/_software_selection.pm:24 called testapi::check_screen
    <<< check_screen(mustmatch='anaconda_kde_highlighted', timeout=1)
    MATCH(anaconda_kde_highlighted:0.00)
    MATCH(anaconda_kde_highlighted-freetype262:0.00)
    no match 1
    MATCH(anaconda_kde_highlighted:0.00)
    MATCH(anaconda_kde_highlighted-freetype262:0.00)
    no match 0
    MATCH(anaconda_kde_highlighted:0.00)
    MATCH(anaconda_kde_highlighted-freetype262:0.00)
    >>> assert_screen: match=anaconda_kde_highlighted timed out after 1
    Debug: /var/lib/openqa/share/tests/fedora/tests/_software_selection.pm:24 called testapi::send_key
    <<< send_key(key='down')
    Debug: /var/lib/openqa/share/tests/fedora/tests/_software_selection.pm:24 called testapi::check_screen
    <<< check_screen(mustmatch='anaconda_kde_highlighted', timeout=1)
    MATCH(anaconda_kde_highlighted:0.00)
    MATCH(anaconda_kde_highlighted-freetype262:0.00)
    no match 1
    MATCH(anaconda_kde_highlighted:0.00)
    MATCH(anaconda_kde_highlighted-freetype262:0.72)
    no match 0
    MATCH(anaconda_kde_highlighted:0.00)
    MATCH(anaconda_kde_highlighted-freetype262:0.72)

**Comment by okurz** [comment]
did you confirm this is not happening with os-autoinst < 4.3 ?

BTW, there is `send_key_until_needlematch` which should do what you want to achieve in the test code more easily. E.g. see here https://github.com/os-autoinst/os-autoinst-distri-opensuse/search?utf8=%E2%9C%93&q=send_key_until_needlematch

If you observe the issue just in this special case then this might be a feasible and helpful workaround. Then your code could look as easy as:
```
send_key_until_needlematch "anaconda_".$packageset."_highlighted", 'down';
```
which is ~~faster (no 1 second timeout on check_screen), more reliable (if SUT is slowed down by test environment and the effect only happens after the 1 second timeout of check_screen),~~ easier to read.  "send_key_until_needlematch" does by default use a limit of 20 key presses, just as you do, which can be overridden with the third argument. I assumed to do something more fancy than you do but actually it is equivalent - hence the strikethrough - except for the behaviour on not finding a match within the counter limit.

**Comment by AdamWill** [comment]
I didn't re-install 4.2 and double check, but I've had to re-take the needles several times (GNOME people keep fiddling with Cantarell and GTK+...) and I'm pretty sure it worked OK then.

Thanks for the `send_key_until_needlematch` tip, that's helpful!

**Comment by coolo** [comment]
this has always been a problem with interactive mode. The problem is that your check_screen times out while you're creating the needle

**Comment by AdamWill** [comment]
Well, this happens before you get to the point of creating a needle. I can just kick off the test and sit there watching it until it gets to the loop, then *immediately* click 'next', and it doesn't hit the 'down' key. It re-does the 'does this needle match?' check, but it doesn't hit the down key, so it just keeps checking the same screen. I never hit the needle edit interface during this sequence.

**Comment by nilxam** [comment]
sounds it's a issue I was working on, the fix should part of  https://github.com/os-autoinst/os-autoinst/pull/439 ie. continuation in interactive mode is continuing in a loop.

**Comment by okurz** [comment]
@AdamWill is this fixed for you with #439 merged?

**Comment by nilxam** [comment]
and https://github.com/os-autoinst/os-autoinst/pull/450

**Comment by AdamWill** [comment]
Dunno, I've been on vacation. Will try and check soon.

**Comment by AdamWill** [comment]
sorry, still didn't get back to this :( i'm awful, I know.

**Comment by AdamWill** [comment]
OK, just re-tested. It doesn't have this bug any more, but I'm not sure if the new behaviour is exactly as intended either. We still have the same loop in the code. What happens now is it reaches the software selection screen, fails the needle match, and pauses. OK, fine. I hit the 'play' icon...and it now just runs through the loop multiple times until it passes, just like in non-interactive mode. It doesn't go one loop iteration at a time, fail the match, and wait for me to hit 'play' again. Which is kinda what I'd expect? Or at least, that's how it used to work, I think.

**Comment by okurz** [comment]
The interactive mode is an undocumented hack and my current take on this is that we need to describe expected behaviour - at best as a test (runnable specification) - and then check our current implementation against it.

Current limitation regarding testing is probably that phantomjs on any live view will make the process hang in updating the view and you can't interact with the system anymore and also we are not running any qemu instance within our tests of openQA. My current approach would be to test with openQA-within-openQA, e.g. see [here](https://openqa.opensuse.org/tests/215887) for a one-shot testrun of openQA on Tumbleweed (failing because it can't reach openqa.opensuse.org from within the qemu instance).

Independant of the test IMHO we should write down specification as human language, e.g. using BDD templates (Given…When…Then) in either the docs within this repo or on https://progress.opensuse.org/projects/openqav3/wiki/Wiki
So if *you* want to document how the system behaves (or should behave) go ahead.

**Comment by nilxam** [comment]
If you check GettingStarted document, you can see the behavior description there, the content is not perfect but it's documented the correct behavior. At least it was matched what I did/fixed after PR#450. If interactive mode was rewritten already(I don't know is it happened or not) then maybe it's another bug.

**Comment by coolo** [comment]
No, the code adam is testing is after pr#450 - and he says it behaves different than before, but still not proper. #524 will be more brutal to that. But I'm trying the scenario adam describes with it - let's see

**Comment by coolo** [comment]
With 524 it works as expected. It stops after every check_screen.

**Comment by AdamWill** [comment]
Yup, confirmed (although I think it's rather https://github.com/os-autoinst/openQA/pull/767/commits/66b7375a1ab8dc2f1b85647116ef06ec441b71db that fixes it than anything from #524). I've applied both https://github.com/os-autoinst/openQA/pull/767 and https://github.com/os-autoinst/os-autoinst/pull/524 to Fedora's staging instance since they fix several bugs that were annoying me, we'll see how it shakes out in a few days of use.
